python/mspasspy/db/database.py

```diff
@@ -2761,6 +2769,336 @@ def read_ensemble_data(
+                    mspass_object = TimeSeries(md)
+                    ensemble.member.append(mspass_object)
+                else:
+                    # api mismatch here.  This ccore Seismogram constructor
+                    # had an ancestor that had an option to read data here.
+                    # we never do that here
+                    mspass_object = Seismogram(md, False)
+                    ensemble.member.append(mspass_object)
+            except MsPASSError as merr:
+                # if the constructor fails mspass_object will be invalid
+                # To preserve the error we have to create a shell to hold the error
+                if object_type is TimeSeries:
+                    mspass_object = TimeSeries()
+                else:
+                    mspass_object = Seismogram()
+                # Default constructors leaves result marked dead so below should work
+                mspass_object.elog.log_error(merr)
+                return mspass_object
+
+                # not continue step 2 & 3 if the mspass object is dead
+            if is_dead:
+                mspass_object.kill()
+                for msg in log_error_msg:
+                    mspass_object.elog.log_error(
+                        "read_data", msg, ErrorSeverity.Invalid
+                    )
+            else:
+                # 2.load data from different modes
+                mspass_object.live = True
+                # 3.load history
+                if load_history:
+                    history_obj_id_name = (
+                        self.database_schema.default_name("history_object") + "_id"
+                    )
+                    if history_obj_id_name in object_doc:
+                        self._load_history(
+                            mspass_object,
+                            object_doc[history_obj_id_name],
+                            alg_name,
+                            alg_id,
+                            define_as_raw,
+                            retrieve_history_record,
+                        )
+
+                mspass_object.clear_modified()
+
+                # 4.post complaint elog entries if any
+                for msg in log_error_msg:
+                    mspass_object.elog.log_error(
+                        "read_data", msg, ErrorSeverity.Complaint
+                    )
+
+        res = cur_collection.aggregate(
+            [
+                {"$match": {"$expr": {"$in": ["$_id", objectid_list]}}},
+                {
+                    "$group": {
+                        "_id": {"dir": "$dir", "dfile": "$dfile"},
+                        "foffs": {"$push": "$foff"},
+                        "ids": {"$push": "$_id"},
+                    }
+                },
+            ]
+        )
+
+        files = list(res)
+        for f in files:
+            cur_dir = f["_id"]["dir"]
+            cur_dfile = f["_id"]["dfile"]
+            foofs = list(map(int, f["foffs"]))  # string list -> int list
```

Comment:
This section needs changes to match what I suggested for the C++ function being used here.  

Also you need to add some comment here about how the python code interacts with the C code and what assumptions both sides make for efficiency.   For your education that is always good practice but especially for a mixed C - python example like this.   We all get lazy and violate that rule way too often.